site/en/docs/extensions/mv3/end-to-end-testing/index.md

```diff
@@ -0,0 +1,98 @@
+title: "End-to-end testing Chrome Extensions"
+seoTitle: "End-to-end testing Chrome Extensions"
+date: 2023-10-09
+description: How to write end-to-end tests for extensions.
+---
+
+End-to-end testing involves an extension package being built and loaded into a browser. A testing
+tool communicates with the browser to automate interactions and test the same flows that a user
+would go through. A library that supports end-to-end testing will generally provide ways of
+controlling the browser, simulating user input and observing the current state of any open pages.
+
+See [Testing Chrome Extensions with Puppeteer][tutorial] for a tutorial.
+
+## Using browser testing libraries
+
+Extensions are supported by a range of testing libraries.
+
+| Library                | Guidance                                                                                                                              |
+|------------------------|:-------------------------------------------------------------------------------------------------------------------------------------:|
+| Puppeteer / Playwright | See Chrome Extensions ([Puppeteer][puppeteer-testing] / [Playwright][playwright-testing]).                                            |
+| Selenium               | Use the [ChromeOptions][selenium-chromeoptions] object to load extensions. More information is available [here][selenium-extensions]. |
+| WebDriverIO            | See [Web Extension Testing][webdriverio-testing].                                                                                     |
+
+## Running tests in headless Chrome
+
+When running tests as part of an automated workflow, it is often necessary to load your extension on a machine that does not have a screen. Chrome's [new headless][new-headless] mode allows Chrome to be run in an unattended environment like this. Simply start Chrome with the `--headless=new` flag (headless currently defaults to "old", which does not support loading extensions). Depending on the automation tool you choose there may be a setting which adds the flag for you automatically.
+
+## Setting an extension ID
+
+It is often desirable to have a fixed extension ID in tests. This makes a number of common tasks easier such as allow-listing the extension's origin on a server it needs to communicate with, or opening extension pages within tests. To do this, follow the steps under [Keeping a consistent extension ID][consistent-id].
+
+## Testing extension pages
+
+Extension pages can be accessed using their corresponding URL, e.g `chrome-extension://<id>/index.html`. Use the normal navigation methods available in your automation tool of choice to navigate to these URLs.
+
+## Testing an extension popup
+
+Opening an extension popup in the context of another page is not currently possible. Instead, open the popup's URL in a new tab. If your popup uses the active tab, consider implementing an override where a tab ID can be passed explicitly to your popup. For example:
+
```

Comment:
Can we add a filename label here too?

Reply:
I added labels to all of the tutorial and places where the filename is important, but I don't think it makes sense here? This code would work anywhere and depends entirely on your project.